### PR TITLE
Feature/timeshift validation

### LINKED ISSUE #460: Validate if time shifted periods intersects
In TimeFilterMenu check if periods after time shift intersects. For example for "Last week" and time shift "1 Day", periods would intersect.

Such comparisons are hard to read and could lead to possible errors in query.

Originaly found in #449

### PR DESCRIPTION
**Pull request by adrianmroz**
Closes #460 

Checks if time shift is to small when creating Essence (from ViewDefinition) and uses empty time shift instead.

In FilterMenu shows error and disables modal if periods overlaps due to small time shift

### CODE REVIEW

**Comment by adrianmroz** [comment]
Need better copy and style for overlap error.

File: src/client/components/filter-menu/boolean-filter-menu/boolean-filter-menu.scss [hunk 1]
```diff
@@ -25,6 +25,7 @@ $checkbox-width: 20px;
     @extend %white-shadow;
     background: $white;
     color: $text-standard;
+    height: 80px;
```
Comment (bigsergey) [inline src/client/components/filter-menu/boolean-filter-menu/boolean-filter-menu.scss]: some values are variables (colors for example), maybe height should be variable too?

File: src/client/components/filter-menu/boolean-filter-menu/boolean-filter-menu.tsx [hunk 2]
```diff
@@ -105,10 +105,11 @@ export class BooleanFilterMenu extends React.Component<BooleanFilterMenuProps, B
 
   }
 
-  constructFilter(): Filter {
-    const { essence: { filter }, dimension } = this.props;
+  constructFilter(): Filter | null {
     const { selectedValues } = this.state;
+    if (selectedValues.isEmpty()) return null;
```
Comment (bigsergey) [inline src/client/components/filter-menu/boolean-filter-menu/boolean-filter-menu.tsx]: tests?

File: src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx [hunk 3]
```diff
@@ -74,16 +75,57 @@ export class PresetTimeTab extends React.Component<PresetTimeTabProps, PresetTim
...
+    return filterPeriod && this.isDurationValid() && this.isTimeShiftValid() && !this.doesTimeShiftOverlap();
+  }
+
+  isFilterDifferent(): boolean {
+    const { essence: { filter, timeShift } } = this.props;
+    const newTimeShift = this.constructTimeShift();
+    const newFilter = this.constructRelativeFilter();
+    return !filter.equals(newFilter) || !timeShift.equals(newTimeShift);
+  }
+
```
Comment (bigsergey) [inline src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx]: Some of the added functions look like util function with can be outside react component. Separate file can be better than class methods? Unit tests will be easer to write.

Comment (adrianmroz) [reply src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx]: Do you think such unit tests would be meaningful? I'd like to write e2e for checking these behaviours. 

And if you extract methods as functions you would need to pass everywhere whole state and props. I don't think readability would increase much.

File: src/client/components/filter-menu/boolean-filter-menu/boolean-filter-menu.tsx [hunk 4]
```diff
@@ -105,10 +105,11 @@ export class BooleanFilterMenu extends React.Component<BooleanFilterMenuProps, B
 
   }
 
-  constructFilter(): Filter {
-    const { essence: { filter }, dimension } = this.props;
+  constructFilter(): Filter | null {
```
Comment (adrianmroz) [inline src/client/components/filter-menu/boolean-filter-menu/boolean-filter-menu.tsx]: e2e test

File: src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx [hunk 5]
```diff
@@ -23,7 +23,7 @@ import { Dimension } from "../../../../common/models/dimension/dimension";
 import { Essence } from "../../../../common/models/essence/essence";
```
Comment (adrianmroz) [inline src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx]: e2e tests for validation (messages and disable/enable state for action buttons)

File: src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx [hunk 6]
```diff
@@ -20,7 +20,8 @@ import { Clicker } from "../../../../common/models/clicker/clicker";
 import { Dimension } from "../../../../common/models/dimension/dimension";
```
Comment (adrianmroz) [inline src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx]: e2e tests for validation (messages and disable/enable state for action buttons)

File: src/common/models/date-range/date-range.ts [hunk 7]
```diff
@@ -24,4 +26,13 @@ interface DateRangeDefinition {
 const defaultDateRange: DateRangeDefinition = { start: null, end: null };
 
 export class DateRange extends Record<DateRangeDefinition>(defaultDateRange) {
+  intersects(other: DateRange | null): boolean {
```
Comment (adrianmroz) [inline src/common/models/date-range/date-range.ts]: unit tests

File: src/common/models/essence/essence.ts [hunk 8]
```diff
@@ -211,15 +211,23 @@ export class Essence extends ImmutableRecord<EssenceValue>(defaultEssence) {
     const isPinnedSortValid = series && constrainedSeries.hasMeasureSeries(pinnedSort);
     const constrainedPinnedSort = isPinnedSortValid ? pinnedSort : Essence.defaultSortReference(constrainedSeries, dataCube);
 
+    const constrainedFilter = filter && filter.constrainToDimensions(dataCube.dimensions);
+
+    const validTimezone = timezone || Timezone.UTC;
+
+    const timeFilter = filter && filter.getClauseForDimension(dataCube.getTimeDimension());
+    if (!(timeFilter instanceof FixedTimeFilterClause || timeFilter instanceof RelativeTimeFilterClause)) throw new Error(`Unknown time filter: ${timeFilter}`);
+    const constrainedTimeShift = timeShift.constrainToFilter(timeFilter, validTimezone);
```
Comment (adrianmroz) [inline src/common/models/essence/essence.ts]: add test cases in converters (main entry point for this constructor)

Comment (adrianmroz) [inline src/common/models/essence/essence.ts]: Should I add validation for this invariant also in `changeTimeShift`?

File: src/common/models/time-shift/time-shift.ts [hunk 9]
```diff
@@ -83,4 +84,22 @@ export class TimeShift implements Instance<TimeShiftValue, TimeShiftJS> {
...
+        return !range.intersects(range.shift(this.value, timezone));
+      case FilterTypes.RELATIVE_TIME:
+        const { duration } = timeFilter as RelativeTimeFilterClause;
+        return this.value.getCanonicalLength() >= duration.getCanonicalLength();
+      default:
+        throw new Error(`Incorrect time filter: ${timeFilter}`);
+    }
+  }
+
+  constrainToFilter(timeFilter: TimeFilterClause, timezone: Timezone): TimeShift {
```
Comment (adrianmroz) [inline src/common/models/time-shift/time-shift.ts]: unit test

File: src/common/models/essence/essence.ts (hunk 8 above)
Comment (adrianmroz) [reply src/common/models/essence/essence.ts]: and when changing time filter ...

File: src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx [hunk 10]
```diff
@@ -76,35 +76,59 @@ export class FixedTimeTab extends React.Component<FixedTimeTabProps, FixedTimeTa
...
+    const clause = new FixedTimeFilterClause({ reference: name, values: List.of(this.createDateRange()) });
     return filter.setClause(clause);
   }
 
   constructTimeShift(): TimeShift {
     return TimeShift.fromJS(this.state.shift);
   }
 
-  validateTimeShift(): boolean {
+  doesTimeShiftOverlap(): boolean {
```
Comment (mkuthan) [inline src/client/components/filter-menu/time-filter-menu/fixed-time-tab.tsx]: I would extract this function, should be easier to test

File: src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx (hunk 3 above)
Comment (mkuthan) [reply src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx]: unit tests for some basic time operations are better than e2e

File: package.json [hunk 11]
```diff
@@ -134,6 +135,7 @@
     "@types/webpack-env": "1.14.1",
     "babel-loader": "8.0.6",
     "chai": "4.1.2",
+    "chai-datetime": "1.5.0",
```
Comment (mkuthan) [inline package.json]: Do we already have any datetime library in the project? I would avoid external dependencies if possible.

Comment (adrianmroz) [reply package.json]: It's library for date assertion for tests, so I can write `expect(date).to.equalDate(expected)`

File: src/common/models/time-shift/time-shift.ts [hunk 12]
```diff
@@ -83,4 +84,22 @@ export class TimeShift implements Instance<TimeShiftValue, TimeShiftJS> {
...
+    switch (timeFilter.type) {
+      case FilterTypes.FIXED_TIME:
+        const { values } = timeFilter as FixedTimeFilterClause;
+        const range = values.first();
+        return !range.intersects(range.shift(this.value, timezone));
+      case FilterTypes.RELATIVE_TIME:
+        const { duration } = timeFilter as RelativeTimeFilterClause;
+        return this.value.getCanonicalLength() >= duration.getCanonicalLength();
+      default:
+        throw new Error(`Incorrect time filter: ${timeFilter}`);
```
Comment (mkuthan) [inline src/common/models/time-shift/time-shift.ts]: Rather unknown than incorrect :)

File: src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx (hunk 3 above)
Comment (adrianmroz) [reply src/client/components/filter-menu/time-filter-menu/preset-time-tab.tsx]: We need more meaningful proposals than extract to function and unit test, to have some discussion.

Half of this function is composing predicates. I don't see any point in testing them, because space of possible inputs is much bigger than real scenarios. I want to test if UI responds to user actions. 

If I extract some subset of these methods and test them, that's still not whole story. I need to orchestrate props/state on component and check if correct functions are called. If extracted functions are used only there, what's the point? I split tests, test intermediate results and reduce maintainability.

Point taken, that comparing canonical lengts could b reused in Essence class, so maybe little part could be extracted. But again, here I got text time shift, in Essence I have full object. Still half of this method is glue code.

File: src/common/models/time-shift/time-shift.mocha.ts [hunk 13]
```diff
@@ -40,4 +44,89 @@ describe("isValidTimeShift", () => {
...
+      });
+    });
+
+    describe("Relative time filter", () => {
+      describe("Latest period", () => {
+        it("does not touch if shifted period do not overlap with original", () => {
+          const filter = new RelativeTimeFilterClause({
+            reference: "time",
+            period: TimeFilterPeriod.LATEST,
+            duration: Duration.fromJS("P3D")
```
Comment (erykpiast) [inline src/common/models/time-shift/time-shift.mocha.ts]: You have a difference by one day in the previous tests set, maybe for consistency sake, it should be `P1D` here? Or `P2D`, I'm not sure if I understand it correctly. If you need a test for an edge case, add it separately. Such one may be valuable for a fixed time filter as well.

File: src/common/models/time-shift/time-shift.mocha.ts [hunk 14]
```diff
@@ -40,4 +44,89 @@ describe("isValidTimeShift", () => {
...
+          expect(shift.constrainToFilter(filter, Timezone.UTC).equals(TimeShift.empty())).to.be.true;
+        });
+      });
+
+      describe("Previous period", () => {
+        it("does not touch if shifted period do not overlap with original", () => {
+          const filter = new RelativeTimeFilterClause({
+            reference: "time",
+            period: TimeFilterPeriod.PREVIOUS,
+            duration: Duration.fromJS("P3D")
```
Comment (erykpiast) [inline src/common/models/time-shift/time-shift.mocha.ts]: The same concern applies here and below.

File: src/common/models/time-shift/time-shift.ts (hunk 9 above)
Comment (erykpiast) [inline src/common/models/time-shift/time-shift.ts]: Does it make sense to make `Timezone.UTC` a default value here? It'd make tests a bit more readable.

File: src/common/models/time-shift/time-shift.mocha.ts [hunk 15]
```diff
@@ -40,4 +44,89 @@ describe("isValidTimeShift", () => {
...
+          expect(shift.constrainToFilter(filter, Timezone.UTC).equals(shift)).to.be.true;
+        });
+
+        it("returns empty time shift if shifted period overlap with original", () => {
+          const filter = new RelativeTimeFilterClause({
+            reference: "time",
+            period: TimeFilterPeriod.LATEST,
+            duration: Duration.fromJS("P4D")
+          });
+          expect(shift.constrainToFilter(filter, Timezone.UTC).equals(TimeShift.empty())).to.be.true;
```
Comment (erykpiast) [inline src/common/models/time-shift/time-shift.mocha.ts]: I wonder how much custom chai matcher may help.

`expect(shift.constrainToFilter(filter)).to.equalUsingInstanceMethod(TimeShift.empty())`

File: src/common/models/time-shift/time-shift.mocha.ts [hunk 16]
```diff
@@ -40,4 +44,89 @@ describe("isValidTimeShift", () => {
...
+        });
+        expect(shift.constrainToFilter(filter, Timezone.UTC).equals(shift)).to.be.true;
+      });
+
+      it("returns empty time shift if shifted period overlap with original", () => {
+        const filter = new FixedTimeFilterClause({
+          reference: "time",
+          values: List.of(new DateRange({ start: new Date("2010-01-01"), end: new Date("2010-01-04") }))
+        });
+        expect(shift.constrainToFilter(filter, Timezone.UTC).equals(TimeShift.empty())).to.be.true;
```
Comment (erykpiast) [inline src/common/models/time-shift/time-shift.mocha.ts]: Maybe name and reuse the result of `TimeShift.empty()`? I'd be quite a nice parity with `shift`.

File: src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx [hunk 17]
```diff
@@ -77,12 +72,12 @@ export class TimeFilterMenu extends React.Component<TimeFilterMenuProps, TimeFil
     if (!dimension) return null;
     const menuSize = Stage.fromSize(MENU_WIDTH, 410);
 
-    const tabs = [TimeFilterMenu.RELATIVE_TAB, TimeFilterMenu.FIXED_TAB].map(name => {
+    const tabs = [TimeFilterTab.RELATIVE, TimeFilterTab.FIXED].map(name => {
       return {
         isSelected: tab === name,
-        title: (name === TimeFilterMenu.RELATIVE_TAB ? STRINGS.relative : STRINGS.fixed),
+        title: (name === TimeFilterTab.RELATIVE ? STRINGS.relative : STRINGS.fixed),
```
Comment (erykpiast) [inline src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx]: I'd name it in the current scope or extract as a function. `title: getTitle(name)` - how sweet it looks, doesn't it?

File: src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx [hunk 18]
```diff
@@ -43,31 +43,26 @@ export interface TimeFilterMenuProps {
...
+  tab: TimeFilterTab;
 }
 
-export class TimeFilterMenu extends React.Component<TimeFilterMenuProps, TimeFilterMenuState> {
-
-  private static readonly RELATIVE_TAB = "relative";
-  private static readonly FIXED_TAB = "fixed";
-
-  public mounted: boolean;
+function initialTab({ essence: { filter }, dimension: { name } }: TimeFilterMenuProps): TimeFilterTab {
```
Comment (erykpiast) [inline src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx]: This name doesn't look like a function.

File: src/common/models/time-shift/time-shift.mocha.ts (hunk 13 above)
Comment (adrianmroz) [reply src/common/models/time-shift/time-shift.mocha.ts]: I don't follow. It's basically product of each `TimeFilterPeriod` and case for longer and case for shorter duration.

File: src/common/models/time-shift/time-shift.ts (hunk 9 above)
Comment (adrianmroz) [reply src/common/models/time-shift/time-shift.ts]: No. I don't think changing API for test convenience is a good idea.

File: src/common/models/time-shift/time-shift.mocha.ts (hunk 15 above)
Comment (adrianmroz) [reply src/common/models/time-shift/time-shift.mocha.ts]: Yeah, probably would be nice.

File: src/common/models/time-shift/time-shift.mocha.ts (hunk 13 above)
Comment (erykpiast) [reply src/common/models/time-shift/time-shift.mocha.ts]: I supposed this value is similar in some way to the pair of `2010-01-01` and `2010-01-02` used in the test for fixed time filter, that's why the result is the same.

File: src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx (hunk 18 above)
Comment (adrianmroz) [reply src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx]: function names: noun is getter, verb is side-effect

File: src/common/models/time-shift/time-shift.mocha.ts (hunk 13 above)
Comment (adrianmroz) [reply src/common/models/time-shift/time-shift.mocha.ts]: I'm reusing `shift` so I suppose I should reuse real range (relative or fixed). I will unify those cases.

File: src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx (hunk 18 above)
Comment (erykpiast) [reply src/client/components/filter-menu/time-filter-menu/time-filter-menu.tsx]: 😱

File: cypress/integration/relative-time-filter-menu.spec.js [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+  });
+
+  it.only("Time shift smaller than base period shows overlappng error", () => {
+    previousPreset("W").click();
+    timeShiftPreset("D").click();
+
+    overlappingError().should("contain", "Shifted period overlaps with main period");
+  });
+
+  it.only("Time shift smaller than base period disables Ok button", () => {
```
Comment (erykpiast) [inline cypress/integration/relative-time-filter-menu.spec.js]: 😱
